app/routes.py

```diff
@@ -1 +1,275 @@
+
+@task_bp.route("", methods=["GET"])
+def read_all_tasks():
+
+    sort_query = request.args.get("sort", "asc")
+    title_query = request.args.get("title")
+    description_query = request.args.get("description")
+   
+
+    if sort_query == "asc":
+        tasks = Task.query.order_by(Task.title.asc())
+
+    elif sort_query == "desc":
+        tasks = Task.query.order_by(Task.title.desc())
+
+    elif title_query:
+        tasks = Task.query.filter_by(title = title_query)
+
+    elif description_query:
+        tasks = Task.query.filter_by(description = description_query)
+
+    else:
+        tasks = Task.query.all()
+
+    tasks_response = []
+
+    for task in tasks:
+        tasks_response.append(task.to_dict())
+    return jsonify(tasks_response), 200
+   
+
+@task_bp.route("/<id>", methods=["GET"])
+def read_single_task(id):
+    task = validate_model(Task, id)
+
+    goal = Goal.query.get(id)
+
+    if goal != None:
+        task_dict = task.to_dict()
+        task_dict["goal_id"] = goal.goal_id
+
+    else:
+        task_dict = task.to_dict()
+
+    return ({"task":task_dict}),200
+    
+
+@task_bp.route("/<task_id>", methods=["PUT"])
+def update_task(task_id):
+    task = validate_model(Task, task_id)
+
+    request_body = request.get_json()
+
+    task.title  = request_body["title"]
+    task.description = request_body["description"]
+
+    db.session.commit()
+
+    return task.to_dict_one_task(), 200
+
+@task_bp.route("/<task_id>", methods=["DELETE"])
+def delete_task(task_id):
+    task = validate_model(Task, task_id)
+
+    db.session.delete(task)
+    db.session.commit()
+
+    return jsonify({"details":f'Task {task_id} "{task.title}" successfully deleted'}), 200
+
+    
+@task_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def mark_incomplete(task_id):
+    task = validate_model(Task, task_id)
+
+    if task.completed_at is not None:
+        task.completed_at = None
+    
+    task.is_complete = False
+    db.session.commit()
+    return jsonify({"task": task.to_dict()}), 200
+
+
+@task_bp.route("<task_id>/mark_complete", methods=['PATCH'])
+def mark_task_complete_slack(task_id):
+    task = validate_model(Task, task_id)
+    
+    if task.completed_at is not None:
+        return jsonify({"task": task.to_dict()}), 200
+    
+    task.completed_at = datetime.utcnow()
+    db.session.commit()
+
+    SLACK_BOT_TOKEN = os.environ.get('SLACK_BOT_TOKEN')
+    SLACK_CHANNEL = '#task-notifications'
+
+ #respone notification to Slack
+
+    message = f"Someone just completed the task{task.title}"
+    slack_data = {'text': message, 'channel': SLACK_CHANNEL}
+    headers = {'Authorization': f'Bearer {SLACK_BOT_TOKEN}'}
+    
+    try:
+        response = requests.post('https://slack.com/api/chat.postMessage', json=slack_data, headers=headers)
+        response.raise_for_status()
+
+    except requests.exceptions.RequestException as e:
+        return jsonify({'message': 'Failed to send Slack notification: {e}'})
```

Comment:
Nice work with the `try/except` clause and on having your Slack post sent after the logic of marking a task complete! We don't want to send out any false positive alerts just in case our logic fails during the update!